File: stuaidapp/types/recognition.ts
/** 困难认定申请表单类型（字段对齐 backend/internal/dto/recognition.go） */

export interface FamilyMemberInput {
  name: string;
  age: number;
  relation: string;
  work_unit: string;
  occupation: string;
  annual_income: number;
  health: string;
}

export interface RecognitionFormState {
  year: number;

  nation: string;
  native_place: string;
  id_card: string;
  family_population: number;
  phone: string;
  address: string;
  postal_code: string;
  guardian_phone: string;

  household_type: string;
  income_source: string;

  special_types: string[];

  natural_disaster: string;
  sudden_accident: string;
  weak_labor: string;
  unemployment: string;
  debt: string;
  other_info: string;

  commitment_agreed: boolean;

  family_members: FamilyMemberInput[];
}

export function emptyRecognitionForm(): RecognitionFormState {
  return {
    year: new Date().getFullYear(),
    nation: 'han',
    native_place: '',
    id_card: '',
    family_population: 1,
    phone: '',
    address: '',
    postal_code: '',
    guardian_phone: '',
    household_type: '',
    income_source: 'wage',
    special_types: [],
    natural_disaster: '',
    sudden_accident: '',
    weak_labor: '',
    unemployment: '',
    debt: '',
    other_info: '',
    commitment_agreed: false,
    family_members: [],
  };
}

export function emptyFamilyMember(): FamilyMemberInput {
  return {
    name: '',
    age: 0,
    relation: 'father',
    work_unit: '',
    occupation: 'worker',
    annual_income: 0,
    health: 'good',
  };
}

// ===== 认定审核（读取后端详情/列表，与 backend/internal/dto/recognition.go 对齐） =====

export type ApplicationStatus =
  | 'draft'
  | 'pending_class'
  | 'pending_dept'
  | 'pending_college'
  | 'pending_final'
  | 'approved'
  | 'rejected';

export type DifficultyLevel = '' | 'special' | 'hard' | 'general';

export interface RecognitionFamilyMember extends FamilyMemberInput {
  id: number;
  special_type: string;
}

export interface ReviewRecord {
  id: number;
  level: number;
  reviewer_id: number;
  reviewer_name: string;
  action: string;
  opinion: string;
  difficulty_level: string;
  reject_to_level: number;
  created_at: string;
}

export interface RecognitionListItem {
  id: number;
  student_id: number;
  student_no: string;
  student_name: string;
  dept_name: string;
  major_name: string;
  class_name: string;
  year: number;
  status: ApplicationStatus;
  current_level: number;
  difficulty_level: DifficultyLevel;
  per_capita_annual_income: number;
}

export interface RecognitionDetail {
  id: number;
  student_id: number;
  student_no: string;
  student_name: string;
  status: ApplicationStatus;
  current_level: number;
  difficulty_level: DifficultyLevel;
  reject_reason: string;

  year: number;
  nation: string;
  native_place: string;
  id_card: string;
  family_population: number;
  phone: string;
  address: string;
  postal_code: string;
  guardian_phone: string;

  household_type: string;
  per_capita_annual_income: number;
  income_source: string;

  special_types: string[];

  natural_disaster: string;
  sudden_accident: string;
  weak_labor: string;
  unemployment: string;
  debt: string;
  other_info: string;

  commitment_agreed: boolean;

  family_members: RecognitionFamilyMember[];
  reviews: ReviewRecord[];
}

export interface ReviewActionInput {
  difficulty_level?: string;
  opinion?: string;
  reject_to_level?: number;
}

File: backend/internal/dto/recognition.go
package dto

import (
	"strings"

	"github.com/wangyifeng2025/student-aid-system/internal/model"
)

// ===== 家庭成员 =====

// FamilyMemberInput 家庭成员录入项。
type FamilyMemberInput struct {
	Name         string  `json:"name"`
	Age          int     `json:"age"`
	Relation     string  `json:"relation"`   // 字典 relation
	WorkUnit     string  `json:"work_unit"`  // 工作/学习单位
	Occupation   string  `json:"occupation"` // 字典 occupation
	AnnualIncome float64 `json:"annual_income"`
	Health       string  `json:"health"`       // 字典 health_status
	SpecialType  string  `json:"special_type"` // 特殊群体类型（可空）
}

// FamilyMemberResponse 家庭成员响应。
type FamilyMemberResponse struct {
	ID           uint    `json:"id"`
	Name         string  `json:"name"`
	Age          int     `json:"age"`
	Relation     string  `json:"relation"`
	WorkUnit     string  `json:"work_unit"`
	Occupation   string  `json:"occupation"`
	AnnualIncome float64 `json:"annual_income"`
	Health       string  `json:"health"`
	SpecialType  string  `json:"special_type"`
}

// ===== 认定申请 =====

// RecognitionRequest 认定申请创建/修改请求（草稿与提交共用，提交时做完整校验）。
type RecognitionRequest struct {
	Year int `json:"year" binding:"required"`

	// 基本情况
	Nation           string `json:"nation"`
	NativePlace      string `json:"native_place"`
	IDCard           string `json:"id_card"`
	FamilyPopulation int    `json:"family_population"`
	Phone            string `json:"phone"`
	Address          string `json:"address"`
	PostalCode       string `json:"postal_code"`
	GuardianPhone    string `json:"guardian_phone"`

	// 家庭经济情况
	HouseholdType         string  `json:"household_type"`
	PerCapitaAnnualIncome float64 `json:"per_capita_annual_income"`
	IncomeSource          string  `json:"income_source"`

	// 特殊群体勾选
	SpecialTypes []string `json:"special_types"`

	// 影响家庭经济状况信息
	NaturalDisaster string `json:"natural_disaster"`
	SuddenAccident  string `json:"sudden_accident"`
	WeakLabor       string `json:"weak_labor"`
	Unemployment    string `json:"unemployment"`
	Debt            string `json:"debt"`
	OtherInfo       string `json:"other_info"`

	// 个人承诺
	CommitmentAgreed bool `json:"commitment_agreed"`

	FamilyMembers []FamilyMemberInput `json:"family_members"`
}

// RecognitionResponse 认定申请详情响应。
type RecognitionResponse struct {
	ID          uint   `json:"id"`
	StudentID   uint   `json:"student_id"`
	StudentNo   string `json:"student_no"`
	StudentName string `json:"student_name"`
	DeptName    string `json:"dept_name"`
	ClassName   string `json:"class_name"`
	Year        int    `json:"year"`

	Nation           string `json:"nation"`
	NativePlace      string `json:"native_place"`
	IDCard           string `json:"id_card"`
	FamilyPopulation int    `json:"family_population"`
	Phone            string `json:"phone"`
	Address          string `json:"address"`
	PostalCode       string `json:"postal_code"`
	GuardianPhone    string `json:"guardian_phone"`

	HouseholdType         string  `json:"household_type"`
	PerCapitaAnnualIncome float64 `json:"per_capita_annual_income"`
	IncomeSource          string  `json:"income_source"`

	SpecialTypes []string `json:"special_types"`

	NaturalDisaster string `json:"natural_disaster"`
	SuddenAccident  string `json:"sudden_accident"`
	WeakLabor       string `json:"weak_labor"`
	Unemployment    string `json:"unemployment"`
	Debt            string `json:"debt"`
	OtherInfo       string `json:"other_info"`

	CommitmentAgreed bool `json:"commitment_agreed"`

	Status          string `json:"status"`
	CurrentLevel    int    `json:"current_level"`
	DifficultyLevel string `json:"difficulty_level"`
	RejectReason    string `json:"reject_reason"`

	FamilyMembers []FamilyMemberResponse `json:"family_members"`
	Reviews       []ReviewRecordResponse `json:"reviews"`
}

// RecognitionListItem 认定申请列表项（精简）。
type RecognitionListItem struct {
	ID                    uint    `json:"id"`
	StudentID             uint    `json:"student_id"`
	StudentNo             string  `json:"student_no"`
	StudentName           string  `json:"student_name"`
	DeptName              string  `json:"dept_name"`
	MajorName             string  `json:"major_name"`
	ClassName             string  `json:"class_name"`
	Year                  int     `json:"year"`
	Status                string  `json:"status"`
	CurrentLevel          int     `json:"current_level"`
	DifficultyLevel       string  `json:"difficulty_level"`
	PerCapitaAnnualIncome float64 `json:"per_capita_annual_income"`
}

// SubmitResult 提交评审结果：申请快照 + 非阻断性提示（单亲/单薪等）。
type SubmitResult struct {
	Application *RecognitionResponse `json:"application"`
	Warnings    []string             `json:"warnings"`
}

// SplitSpecialTypes 将逗号分隔的特殊群体集合拆为切片（去空）。
func SplitSpecialTypes(s string) []string {
	if strings.TrimSpace(s) == "" {
		return []string{}
	}
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if t := strings.TrimSpace(p); t != "" {
			out = append(out, t)
		}
	}
	return out
}

// JoinSpecialTypes 将切片拼为逗号分隔字符串（去空去重）。
func JoinSpecialTypes(types []string) string {
	seen := make(map[string]struct{}, len(types))
	out := make([]string, 0, len(types))
	for _, t := range types {
		t = strings.TrimSpace(t)
		if t == "" {
			continue
		}
		if _, ok := seen[t]; ok {
			continue
		}
		seen[t] = struct{}{}
		out = append(out, t)
	}
	return strings.Join(out, ",")
}

func ToFamilyMemberResponses(items []model.FamilyMember) []FamilyMemberResponse {
	out := make([]FamilyMemberResponse, 0, len(items))
	for i := range items {
		m := &items[i]
		out = append(out, FamilyMemberResponse{
			ID:           m.ID,
			Name:         m.Name,
			Age:          m.Age,
			Relation:     m.Relation,
			WorkUnit:     m.WorkUnit,
			Occupation:   m.Occupation,
			AnnualIncome: m.AnnualIncome,
			Health:       m.Health,
			SpecialType:  m.SpecialType,
		})
	}
	return out
}

// ToRecognitionResponse 组装详情响应；studentNo/studentName 由调用方传入（来自关联学生）。
func ToRecognitionResponse(a *model.RecognitionApplication, studentNo, studentName string) RecognitionResponse {
	return RecognitionResponse{
		ID:                    a.ID,
		StudentID:             a.StudentID,
		StudentNo:             studentNo,
		StudentName:           studentName,
		Year:                  a.Year,
		Nation:                a.Nation,
		NativePlace:           a.NativePlace,
		IDCard:                a.IDCard,
		FamilyPopulation:      a.FamilyPopulation,
		Phone:                 a.Phone,
		Address:               a.Address,
		PostalCode:            a.PostalCode,
		GuardianPhone:         a.GuardianPhone,
		HouseholdType:         string(a.HouseholdType),
		PerCapitaAnnualIncome: a.PerCapitaAnnualIncome,
		IncomeSource:          a.IncomeSource,
		SpecialTypes:          SplitSpecialTypes(a.SpecialTypes),
		NaturalDisaster:       a.NaturalDisaster,
		SuddenAccident:        a.SuddenAccident,
		WeakLabor:             a.WeakLabor,
		Unemployment:          a.Unemployment,
		Debt:                  a.Debt,
		OtherInfo:             a.OtherInfo,
		CommitmentAgreed:      a.CommitmentAgreed,
		Status:                string(a.Status),
		CurrentLevel:          int(a.CurrentLevel),
		DifficultyLevel:       string(a.DifficultyLevel),
		RejectReason:          a.RejectReason,
		FamilyMembers:         ToFamilyMemberResponses(a.FamilyMembers),
		Reviews:               []ReviewRecordResponse{},
	}
}

// ===== 附件 =====

// AttachmentResponse 附件响应。
type AttachmentResponse struct {
	ID         uint   `json:"id"`
	OwnerType  string `json:"owner_type"`
	OwnerID    uint   `json:"owner_id"`
	FileName   string `json:"file_name"`
	Size       int64  `json:"size"`
	Mime       string `json:"mime"`
	UploaderID uint   `json:"uploader_id"`
}

func ToAttachmentResponse(a *model.Attachment) AttachmentResponse {
	return AttachmentResponse{
		ID:         a.ID,
		OwnerType:  a.OwnerType,
		OwnerID:    a.OwnerID,
		FileName:   a.FileName,
		Size:       a.Size,
		Mime:       a.Mime,
		UploaderID: a.UploaderID,
	}
}

func ToAttachmentResponses(items []model.Attachment) []AttachmentResponse {
	out := make([]AttachmentResponse, 0, len(items))
	for i := range items {
		out = append(out, ToAttachmentResponse(&items[i]))
	}
	return out
}
